File: leetcode/CoinChange.py
from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        """
        Calculates the fewest number of coins needed to make up a given amount.
        This is solved using bottom-up dynamic programming.
        """
        # dp[i] will store the minimum number of coins to make amount i.
        # Initialize with a value representing infinity. `float('inf')` is a
        # clear way to represent an unreachable state.
        dp = [float('inf')] * (amount + 1)

        # Base case: 0 coins are needed to make an amount of 0.
        dp[0] = 0

        # Sorting the coins allows for a small optimization in the inner loop.
        coins.sort()

        # Iterate through all amounts from 1 to `amount` to fill the DP table.
        for i in range(1, amount + 1):
            # For each amount, try every coin.
            for coin in coins:
                # If the current coin is larger than the target amount `i`,
                # all subsequent coins will also be larger, so we can stop checking.
                if coin > i:
                    break
                # Update the DP table for amount `i`. The number of coins is
                # 1 (the current coin) + the minimum coins for the remaining amount.
                dp[i] = min(dp[i], 1 + dp[i - coin])

        # If dp[amount] is still infinity, it means the amount cannot be formed.
        result = dp[amount]
        return int(result) if result != float('inf') else -1

File: leetcode/medium/CoinChange.java
import java.util.Arrays;

/**
 * Solution for the Coin Change problem using Dynamic Programming.
 */
class Solution {

    /**
     * Calculates the fewest number of coins needed to make up a given amount.
     *
     * @param coins  An array of integers representing the coin denominations.
     * @param amount The total amount of money.
     * @return The fewest number of coins needed, or -1 if the amount cannot be made.
     */
    public int coinChange(int[] coins, int amount) {
        // Optimization: Sort coins to enable early breaking in the inner loop.
        Arrays.sort(coins);

        // dp[i] stores the minimum coins needed for amount i.
        // Size is amount + 1 because we need to store data for 0 to amount.
        int[] dp = new int[amount + 1];

        // Initialize dp array with a value larger than any possible solution.
        // amount + 1 is effectively "infinity" in this context because
        // even with coins of value 1, the max coins needed is 'amount'.
        Arrays.fill(dp, amount + 1);

        // Base case: 0 coins are needed to make amount 0.
        dp[0] = 0;

        // Iterate through every amount from 1 to the target amount.
        for (int i = 1; i <= amount; i++) {
            // Try every coin denomination
            for (int coin : coins) {
                // If the current coin is larger than the current amount i,
                // we cannot use this coin or any subsequent coins (since it's sorted).
                if (coin > i) {
                    break;
                }

                // Recurrence relation:
                // dp[i] = min(current value, 1 coin + result for remainder amount)
                dp[i] = Math.min(dp[i], dp[i - coin] + 1);
            }
        }

        // Check if the value at dp[amount] is still the initialized "infinity".
        // If so, it means the amount could not be reached.
        return dp[amount] > amount ? -1 : dp[amount];
    }
}